torch/distributed/fsdp/_shard_utils.py

```diff
@@ -150,16 +154,39 @@ def _all_gather_sharded_tensor(
         pg = distributed_c10d._get_default_group()
     world_size = dist.get_world_size(pg)
     shards = sharded_tensor.local_shards()
-    local_tensor = shards[0].tensor.flatten()
     dim_0_size = sharded_tensor.size()[0]  # type: ignore[index]
     tensor_numel = sharded_tensor.size().numel()  # type: ignore[union-attr]
     chunk_size = math.ceil(dim_0_size / world_size) * tensor_numel // dim_0_size
-    num_padding = chunk_size - local_tensor.numel()
-    if num_padding > 0:
-        local_tensor = F.pad(local_tensor, [0, num_padding])
-    tensor = torch.empty(chunk_size * world_size, dtype=local_tensor.dtype).cuda()
+    cuda_device = torch.device("cuda", torch.cuda.current_device())
+    if shards:
+        local_tensor = shards[0].tensor.flatten()
+        if not local_tensor.is_cuda:
+            move_to_cpu = torch.ones(1, device=cuda_device)
+            local_tensor = local_tensor.cuda()
+        else:
+            move_to_cpu = torch.zeros(1, device=cuda_device)
+        num_padding = chunk_size - local_tensor.numel()
+        if num_padding > 0:
+            local_tensor = F.pad(local_tensor, [0, num_padding])
+    else:
+        local_tensor = torch.zeros(
+            chunk_size, dtype=sharded_tensor.dtype, device=cuda_device
+        )
+        move_to_cpu = torch.zeros(1, device=cuda_device)
+
+    tensor = torch.empty(
+        chunk_size * world_size,
+        dtype=local_tensor.dtype,
+        device=cuda_device,
+    )
     dist._all_gather_base(tensor, local_tensor, group=pg)
-    return tensor.narrow(0, 0, tensor_numel).reshape(sharded_tensor.size())
+    all_move_to_cpu = torch.empty(world_size, device=cuda_device)
+    dist._all_gather_base(all_move_to_cpu, move_to_cpu, group=pg)
+
+    tensor = tensor.narrow(0, 0, tensor_numel).reshape(sharded_tensor.size())
+    if all_move_to_cpu.sum().is_nonzero():
+        tensor = tensor.cpu()
```

Comment:
just realized that maybe we should not move the all gathered data to cpu here, otherwise all ranks are moving redundant all gathered data to cpu and could make cpu OOM. instead, we may be able to move the data back to cpu after sharding as needed.

Reply:
Removed the "moving back to CPU" part and let users to handle it.